### ISSUE TITLE
Mypy should raise error on `Type[T]` where `T` is not valid

### ISSUE THREAD

**Issue by jhance** [python/mypy#4819]
As specified by PEP 484 (at least according  to @ilevkivskyi), the only valid parameters for `Type[]` are proper classes and unions of proper classes.

Something like `Type[Optional[int]]` should raise an error.

**Comment by ilevkivskyi** [comment]
Thanks for reporting!

It looks like this doesn't give an error neither with `--strict-optional` option nor without it.

**Comment by elazarg** [comment]
This could be type-checked inside mypy too, which may help reasoning about certain parts in the code.

**Comment by gvanrossum** [comment]
But why should it be an error? None represents a proper class (NoneType)
and Optional is just a shorthand for a union.

**Comment by ethanhs** [comment]
Theoretically, you should have a `Union[Type[int], Type[None]]`, not `Type[Union[int, None]]`. (This may also have practical repercussions, but I haven't looked at that).

**Comment by gvanrossum** [comment]
Jukka quotes Ivan quoting PEP 484 as saying "the only valid parameters for `Type[]` are proper classes and unions of proper classes." Indeed the PEP has these words:

> Note that it is legal to use a union of classes as the parameter for `Type[]`, as in:
> ```py
>   def new_non_team_user(user_class: Type[Union[BasicUser, ProUser]]):
>       user = new_user(user_class)
>       ...
> ```

But if that's legal, then `Type[Union[int, NoneType]]` should also be legal, and we normally spell that as `Type[Union[int, None]]` which is equivalent to `Type[Optional[int]]`. So I still don't see why Jukka wants that to become an error. (Regardless of whether it's equivalent to `Union[Type[int], Type[None]]`.)

**Comment by ilevkivskyi** [comment]
@gvanrossum 
> None represents a proper class (NoneType)

At runtime yes, but not in mypy (and I believe in any other static type checker). `None` is special in several aspects like being a singleton, non-subclassable, etc. This is why we have `NoneTyp` instead of `TypeInfo` for it. This means that if we are going to allow `Type[None]`, we will need to patch dozens of places in mypy for negligible benefits (if not negative, see below).

But I think a more important argument is that `Type[Optional[C]]` is most likely a mistake, where a user wanted `Optional[Type[C]]` (it is hard to imagine a use case where someone _wants_ to accept `NoneType` as an argument to a function).

Coming back to the question of the PEP wording, there are some things that are proper classes at runtime, but this doesn't make them valid types, for example not so long ago all special forms in `typing` were classes (I guess this is one of the reasons why PEP uses the word "proper"). `None` is much more closer to being a special `typing` form, than to being a proper class. I would actually propose to make this clear in the PEP unless someone disagrees.

**Comment by gvanrossum** [comment]
OK, the PEP is indeed somewhat ambiguous -- first it seems to list `None` as a special construct (with `Any`, `Union` etc.), but in the next section ("Using None") it says explicitly

> When used in a type hint, the expression `None` is considered
> equivalent to `type(None)`.

There's another mention of `type(None)`, in the summary of the `typing` module towards the end:

> * Optional, defined by ``Optional[t] == Union[t, type(None)]``

I'm happy to change these to make it clear that you *must* use `None` (and not `type(None)`) in the PEP, and to clarify that `None` is a special form, not a "proper class".
